File: determination.py
from policy import resolve_policy


def determine(fixture: dict, corpus_dir: str) -> dict:
    x = fixture["input"]
    p = resolve_policy(x["policy"]["version"], x["policy"]["digest"], corpus_dir)
    if not x["authority_valid"]:
        return {"kind":"NOT_ADMIT","reason_code":"DENIED_AUTHORITY"}
    if x["amount"] > p["limits"]["transfer_max"]:
        return {"kind":"NOT_ADMIT","reason_code":"DENIED_POLICY_LIMIT"}
    if x["balance"] < x["amount"]:
        return {"kind":"NOT_ADMIT","reason_code":"DENIED_INSUFFICIENT_FUNDS"}
    if x["risk"] == "elevated":
        return {"kind":"NOT_ADMIT","reason_code":"DENIED_RISK"}
    return {"kind":"ADMIT","reason_code":"APPROVED"}

File: reference-implementation/src/determination.rs
use serde_json::{Value, json};

pub fn determine(fixture: &Value, policy: &Value) -> Value {
    let x=&fixture["input"];
    if !x["authority_valid"].as_bool().unwrap_or(false) { return json!({"kind":"NOT_ADMIT","reason_code":"DENIED_AUTHORITY"}); }
    if x["amount"].as_i64().unwrap_or(0) > policy["limits"]["transfer_max"].as_i64().unwrap_or(-1) { return json!({"kind":"NOT_ADMIT","reason_code":"DENIED_POLICY_LIMIT"}); }
    if x["balance"].as_i64().unwrap_or(0) < x["amount"].as_i64().unwrap_or(0) { return json!({"kind":"NOT_ADMIT","reason_code":"DENIED_INSUFFICIENT_FUNDS"}); }
    if x["risk"].as_str() == Some("elevated") { return json!({"kind":"NOT_ADMIT","reason_code":"DENIED_RISK"}); }
    json!({"kind":"ADMIT","reason_code":"APPROVED"})
}
